service/controller/v5/resource/instance/deployment.go

```diff
@@ -2,17 +2,49 @@ package instance
 
 import (
 	"context"
+	"encoding/base64"
+	"fmt"
 
 	azureresource "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
 	"github.com/Azure/go-autorest/autorest/to"
 	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
 	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
 
+	"github.com/giantswarm/azure-operator/service/controller/v5/blobclient"
 	"github.com/giantswarm/azure-operator/service/controller/v5/controllercontext"
 	"github.com/giantswarm/azure-operator/service/controller/v5/key"
+	"github.com/giantswarm/azure-operator/service/controller/v5/templates"
 )
 
 func (r Resource) newDeployment(ctx context.Context, obj providerv1alpha1.AzureConfig, overwrites map[string]interface{}) (azureresource.Deployment, error) {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return azureresource.Deployment{}, microerror.Mask(err)
+	}
+	err = cc.Validate()
+	if err != nil {
+		return azureresource.Deployment{}, microerror.Mask(err)
+	}
+
+	cloudConfigURLs := []string{
+		key.BlobName(obj, key.PrefixMaster()),
+		key.BlobName(obj, key.PrefixWorker()),
+	}
+
+	for _, key := range cloudConfigURLs {
+		blobExists, err := blobclient.BlobExists(ctx, key, cc.ContainerURL)
+		if err != nil {
+			return azureresource.Deployment{}, microerror.Mask(err)
+		}
+		if !blobExists {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("cloudconfig blob %#q not found", key))
+			resourcecanceledcontext.SetCanceled(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
```

Comment:
IMO second level methods used in the resource interface implementation should not know about control flow primitives of the resource implementation specifics. Here we should rather return an error, which is asserted by the caller. The caller here is either `EnsureCreated` or `EnsureDeleted` or the like. There we can do the resource cancelation. Look at it again from the isolated test perspective. When you want to ensure `newDeployment` behaves as expected, it is way easier to verify the returned error, as compared to no result at all and having to verify the control flow primitives. Further the control flow primitives should stay as close as possible to the resource's interface specifications as mentioned above.

Reply:
I 100% agree with that.

Reply:
That would be a good input to sig-operator repo.